rust/connlib/tunnel/src/io.rs

```diff
@@ -25,16 +30,20 @@ use std::{
 const DNS_QUERIES_QUEUE_SIZE: usize = 100;
 
 /// Bundles together all side-effects that connlib needs to have access to.
-pub struct Io {
+pub struct Io<S: RuntimeSocketProtector> {
```

Comment:
This is quite an invasive abstraction.

Is there are reason why it wasn't possible to retain the previous trait-object based design? Currently, `Sockets` can be configured with a trait-object to protect newly created sockets which we use when we reconnect.

Why can't we do the same thing with the hickory socket? Make a custom struct that stores a trait-object which it applies when it makes a new socket.

`ClientTunnel` should simply take this socket factory as an argument.

To construct one of those factories, we can make platform-cfg'd `new` functions that automatically set the appropriate `protect` callback.

I don't think the current "reuse" of the protect functionality justifies how invasive this abstraction is.